File: Octave/lagrange.m
# Metodo de interpolacion de lagrange
# Entradas: puntos en x y puntos en y
# Salidas: polinomio de lagrange.

function polinomio = lagrange(x_points, y_points)
  pkg load symbolic;
  n = length(x_points);
  x = sym('x');
  polinomio = 0;
  for k =1:n
    prod = 1;
    for i=1:n
      if i != k
        prod = prod*(x - x_points(i)) / (x_points(k) - x_points(i));
      endif
    endfor
    polinomio = polinomio + prod * y_points(k);
  endfor
  polinomio = simplify(polinomio);
endfunction

File: Python/lagrange.py
from sympy import *


"""
Metodo de interpolacion de lagrange
Entradas: puntos en x y puntos en y
Salidas: polinomio de lagrange.
"""
def lagrange(x_points, y_points):
    n = len(x_points)
    x = Symbol('x')
    polinomio = 0
    for k in range(n):
        prod = 1
        for i in range(n):
            if i != k:
                prod *= (x - x_points[i]) / (x_points[k] - x_points[i])
        polinomio += prod * y_points[k]

    return simplify(polinomio)


print(lagrange([-2, 0, 1], [0, 1, -1]))
